File: galgamecommon/src/main/java/com/goexp/galgame/common/website/GetchuURL.java
package com.goexp.galgame.common.website;

import com.goexp.common.util.UrlBuilder;

import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpRequest;
import java.nio.charset.Charset;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;

public final class GetchuURL {

    private final static String searchUrl = "http://www.getchu.com/php/search.phtml";

    public static String getURLfromTitle(String title) {

        return UrlBuilder
                .create(searchUrl)
                .param("sort", "release_date")
                .param("genre", "pc_soft")
                .param("search_keyword", URLEncoder.encode(title, Charset.forName("shift-jis")))
                .build();

    }

    public static String getListByBrand(int brandId) {
        return UrlBuilder
                .create(searchUrl)
                .param("search", "1")
                .param("sort", "release_date")
                .param("genre", "pc_soft")
                .param("search_brand_id", String.valueOf(brandId))
                .param("list_count", "1000")
                .build();
    }

    public static String gameListFromDateRange(LocalDate from, LocalDate to) {
        return UrlBuilder
                .create(searchUrl)
                .param("search", "1")
                .param("sort", "release_date")
                .param("genre", "pc_soft")
                .param("start_date", from.format(DateTimeFormatter.ofPattern("yyyy/MM/dd")))
                .param("end_date", to.format(DateTimeFormatter.ofPattern("yyyy/MM/dd")))
                .param("list_count", "1000")
                .build();
    }

    public static String getFromId(int gameId) {

        return UrlBuilder
                .create("http://www.getchu.com/soft.phtml")
                .param("id", String.valueOf(gameId))
//                .param("gc", "gc")
                .build();

    }


    public static String getImgSmallFromId(int gameId) {

        return String.format("http://www.getchu.com/brandnew/%d/rc%dpackage.jpg", gameId, gameId);
    }

    public static String getImgBigFromId(int gameId) {

        return String.format("http://www.getchu.com/brandnew/%d/c%dpackage.jpg", gameId, gameId);
    }

    public static String getSimpleImgBigFromSrc(String src) {

        return getUrlFromSrc(src);
    }

    public static String getSimpleImgSmallFromSrc(String src) {

        var ts = src.replace(".jpg", "_s.jpg");

        return getUrlFromSrc(ts);
    }

    public static String getUrlFromSrc(String src) {

        if (src.startsWith("http") || src.startsWith("https"))
            return src;

        return String.format("http://www.getchu.com%s", src.replaceFirst("\\.", ""));
    }

//    public static String getCharSmallFromId(int gameId, int charIndex) {
//
//        return String.format("http://www.getchu.com/brandnew/%d/c%dchara%d.jpg", gameId, gameId, charIndex);
//    }

    public static class RequestBuilder {
        private HttpRequest.Builder builder;


        private RequestBuilder(String url) {

            builder = HttpRequest
                    .newBuilder()
                    .uri(URI.create(url))
            ;
        }

        public static RequestBuilder create(String url) {
            return new RequestBuilder(url);
        }


        public RequestBuilder adaltFlag() {
            builder = builder.header("Cookie", "getchu_adalt_flag=getchu.com");
            return this;
        }


        private RequestBuilder webClientParam() {
            builder = builder
                    .header("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8")
                    .header("Accept-Encoding", "gzip, deflate")
                    .header("Accept-Language", "ja,en-US;q=0.7,en;q=0.3")
                    .header("Cache-Control", "max-age=0")
                    .header("Upgrade-Insecure-Requests", "1")
                    .header("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.77 Safari/537.36")
            ;

            return this;
        }


        public HttpRequest build() {

            webClientParam();
            return builder.GET().build();
        }


    }

}

File: galgamecommon/src/main/scala/com/goexp/galgame/common/website/GetchuURL.scala
package com.goexp.galgame.common.website

import java.net.http.HttpRequest
import java.time.LocalDate

import com.goexp.common.util.date.formatDate
import com.goexp.common.util.web.url.{URLSupport, UrlBuilder}

object GetchuURL {


  object GameList {
    def byDateRange(from: LocalDate, to: LocalDate): String =

      UrlBuilder(searchUrl)
        .param("search", 1)
        .param("sort", "release_date")
        .param("genre", "pc_soft")
        .param("start_date", from.format("yyyy/MM/dd"))
        .param("end_date", to.format("yyyy/MM/dd"))
        .param("list_count", 1000)
        .build

    def byBrand(brandId: Int): String =
      UrlBuilder(searchUrl)
        .param("search", 1)
        .param("sort", "release_date")
        .param("genre", "pc_soft")
        .param("search_brand_id", brandId)
        .param("list_count", 1000)
        .build

    private val searchUrl = "http://www.getchu.com/php/search.phtml"

    def getURLfromTitle(title: String): String =
      UrlBuilder(searchUrl)
        .param("sort", "release_date")
        .param("genre", "pc_soft")
        .param("search_keyword", title.urlEncode("shift-jis"))
        .build
  }

  object Game {
    def byId(gameId: Int): String =
      UrlBuilder("http://www.getchu.com/soft.phtml")
        .param("id", gameId)
        .build //                .param("gc", "gc")

    def SmallImg(gameId: Int): String =
      s"http://www.getchu.com/brandnew/$gameId/rc${gameId}package.jpg"

    def LargeImg(gameId: Int): String =
      s"http://www.getchu.com/brandnew/$gameId/c${gameId}package.jpg"

    def largeSimpleImg(src: String): String =
      getUrlFromSrc(src)

    def smallSimpleImg(src: String): String = {
      val ts = src.replace(".jpg", "_s.jpg")
      getUrlFromSrc(ts)
    }

    def getUrlFromSrc(src: String): String = {
      if (src.startsWith("http") || src.startsWith("https")) return src

      s"http://www.getchu.com${src.replaceFirst("\\.", "")}"
    }

    //        public static String charSmall(int gameId, int charIndex) {
    //            return String.format("http://www.getchu.com/brandnew/%d/c%dchara%d.jpg", gameId, gameId, charIndex);
    //        }
  }

  object RequestBuilder {
    def create(url: String) = new RequestBuilder(url)

    def apply(url: String): RequestBuilder = new RequestBuilder(url)
  }

  class RequestBuilder private(val url: String) {


    private var builder = HttpRequest.newBuilder.uri(url.toURI())

    def adaltFlag: RequestBuilder = {
      builder = builder.header("Cookie", "getchu_adalt_flag=getchu.com")
      this
    }

    private def webClientParam = {
      builder = builder
        .header("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8")
        .header("Accept-Encoding", "gzip, deflate")
        .header("Accept-Language", "ja,en-US;q=0.7,en;q=0.3")
        .header("Cache-Control", "max-age=0")
        .header("Upgrade-Insecure-Requests", "1")
        .header("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.77 Safari/537.36")
      this
    }

    def build: HttpRequest = {
      webClientParam
      builder.GET.build
    }
  }

}